Mathlib/Tactic/FailIfNoProgress.lean

```diff
@@ -47,7 +47,8 @@ def lctxIsDefEq : (l₁ l₂ : List (Option LocalDecl)) → MetaM Bool
   | _, _ => return false
 
 /-- Run `tacs : TacticM Unit` on `goal`, and fail if no progress is made. -/
-def runAndFailIfNoProgress (goal : MVarId) (tacs : TacticM Unit) : TacticM (List MVarId) := do
+def runAndFailIfNoProgress (goal : MVarId) (tacs : TacticM Unit) : TacticM (List MVarId) :=
+  goal.withContext do
```

Comment:
This doesn't hurt, but I'm not sure it really fixes the issue. The main problem is that `newGoal` and `goal` might not have compatible local contexts at all, so asking `isDefEq` is senseless.

One strategy that's *almost* the right thing to do is to revert all the local variables in each context, and then do the `isDefEq` check on those. It's not quite right since `isDefEq` will unfold let bindings, so certain kinds of unfoldings won't count as progress.

An alternative is to do the `lctxIsDefEq` check *first*, and then only do the target `isDefEq` if that's satisfied. However, `lctxIsDefEq` does not check defeq of let values, so this is still not correct without a further fix.

Reply:
I agree this is a very minimal fix. But it's already progress in the sense that it fixes a bug in Lean verbose. Feel free to PR a better fix (you can push to this branch).

Reply:
Ok, I'm working on one.

Reply:
Done, assuming CI passes